### DISCUSSION TITLE
想要把【打开文件位置】写一个自定义脚本，给加上快捷键，有人会写脚本吗

### DISCUSSION THREAD

**Discussion by calmwaves111** [windingwind/zotero-actions-tags#307] [Q&A] [answered]
![image](https://github.com/windingwind/zotero-actions-tags/assets/56907077/4bc0291e-da8e-4d52-89d2-97628c5a0041)
就是右键菜单里的这个命令，我希望给它加上个快捷键，想问问怎么用脚本实现这个命令

另外问一问zotero这些api啊函数啊有官方文档吗

**Comment by ramonmi** [comment]
感觉可以参考这个试着改一改 https://github.com/windingwind/zotero-actions-tags/discussions/132

**Comment by calmwaves111** [comment]
> 感觉可以参考这个试着改一改 #132

是的，我就是仿照这个写了脚本

```js
const window = require("window");
let filePath;
  if (item.isRegularItem()) {
    targetItem = (await item.getBestAttachments()).find(
      (att) =>
        att.isPDFAttachment() ||
        att.isSnapshotAttachment() ||
        att.isEPUBAttachment()
    );
    if (targetItem) {
		filePath = await targetItem.getFilePathAsync();
	}
	else{
		Zotero.alert(window, "信息面板", "该条目没有附件");
	}
}
else{
	filePath = await item.getFilePathAsync();
}
let exePath = "D:\\OneCommander\\OneCommander.exe"

Zotero.launchFileWithApplication(filePath, exePath);
return "打开成功";
```

现在的问题是，我使用的是自己安装的文件管理器`D:\\OneCommander\\OneCommander.exe`，我用了一段时间了，这对于我自己来说已经够用了

如果想改成win自带的系统资源管理器，还是有些问题
第一我尝试使用`C:\\Windows\\explorer.exe`，这个问题是，使用的是filePath，所以无法弹出资源管理器窗口，而是直接用默认应用打开文件
第二我尝试获得folderPath，但是使用`Zotero.launchFileWithApplication(folderPath, exePath);`用不了
第三我尝试直接使用命令行`const { exec } = require('child_process');`，但是似乎不支持引入这个模块

希望有大佬能看看帮忙解决

**当然我觉得这个办法还是比较笨的，要自己获取文件路径 ，然后用某个exe打开，我认为zotero既然给右键菜单加了这个【打开文件位置】的选项，我想应该是有api可以直接用的，只是我对此并不熟悉，也没找到文档，就无能为力了**

**Reply by windingwind** [reply]
`Zotero.File.reveal(path)`

**Reply by calmwaves111** [reply]
感谢，这个就默认用设置的文件管理器打开了，不需要设置withApplication
不过这个还是要传入path，我还挺好奇右键菜单【打开文件位置】这个有没有api，因为就算我选中的是条目而不是附件，它也能直接打开附件位置，我以为是有类似于`item.showfile`这样的api

这个功能其实是我想仿zutilo插件，它能把很多右键菜单给设置快捷键，十分可惜的是zutilo不再适配zotero7了

**Reply by windingwind** [reply, marked as answer]
`ZoteroPane.showAttachmentInFilesystem(itemID)`

**Reply by windingwind** [reply]
通常最方便的查询zotero源码对应功能代码的方式是git clone以后搜索你在界面上看到的文本，然后再搜索文本对应的key(在此例中是`locate.showFile.label`)或其部分关键片段（在此例中是`showFile`），然后即可找到对应位置（此例中是`locateMenu.js` > `ViewOptions.showFile` > `handleItems`）

欢迎贡献zutilo脚本的替代

**Comment by calmwaves111** [comment]
用系统资源管理器的可用代码
虽然不优雅，采用的是文件夹路径，不会直接选中文件
可能少一些非空判断
但是总算是实现了，还是希望能找到这个api

```js
const window = require("window");

let filePath;
  if (item.isRegularItem()) {
    targetItem = (await item.getBestAttachments()).find(
      (att) =>
        att.isPDFAttachment() ||
        att.isSnapshotAttachment() ||
        att.isEPUBAttachment()
    );
    if (targetItem) {
		filePath = await targetItem.getFilePathAsync();
	}
	else{
		Zotero.alert(window, "信息面板", "该条目没有附件");
	}
}
else{
	filePath = await item.getFilePathAsync();
}

let folderPath = filePath ? filePath.replace(/\\[^\\]+$/, ''): null;
let exePath = "C:\\Windows\\explorer.exe"

Zotero.launchFileWithApplication(folderPath, exePath);
return "打开成功";
```

**Comment by ProudBenzene** [comment]
综合借助上面作者的回答，补充一段我修改的在macOS和Windows下都可以打开附件文件夹且Zotero不会报错的代码：

```JavaScript
/**
 * Finder Opener
 * @author Calmwave, ProudBenzene
 * @reference  https://github.com/windingwind/zotero-actions-tags/issues/235
 * @usage Open the folder of the attachment in Finder/Default Explorer
 * @link https://github.com/windingwind/zotero-actions-tags/discussions/307
 * @see https://github.com/windingwind/zotero-actions-tags/discussions/307
*/

const win = require ('Zotero').getMainWindow();

targetItem = (await item.getBestAttachments()).find(
	(att) =>
		att.isPDFAttachment() ||
		att.isSnapshotAttachment() ||
		att.isEPUBAttachment()
);

if (!targetItem) {
	return "[Finder Opener] No Attachments in this Item";
}

// Zotero.launchFileWithApplication(folderPath, exePath);
win.ZoteroPane.showAttachmentInFilesystem(targetItem.id);
return "[Finder Opener] 打开成功";
```
